### PR TITLE
Add deployment tests

### PR DESCRIPTION
**Pull request by maggie0002**
Adds a deploy to Cloud test to run after Flowzone completes.

The test:
- Edits the `action.yml` to tell it to point to the docker image built by Flowzone in the build step before. 
- Overwrites the Dockerfile with a simple busy box image so the test completes quicker than trying to deploy the entire deploy-to-balena Dockerfile just for a deploy test
- Pushes to a test fleet to ensure it works.

### CODE REVIEW

**Comment by maggie0002** [comment]
@klutchell, it seems `${{ github.event.pull_request.head.sha }}` isn't available inside the action.yml to specify a container. Here is an alternative way. Bit of a shame to have to have a test action.yml, would rather they used the same file but it's still pretty close.

Ideas welcome. Particularly on whether there is a better way to integrate the test into Flowzone rather than have it sitting below as an additional job (which works fine, but if there is a custom test option for Flowzone, may as well use it as intended).

**Comment by klutchell** [comment]
I was thinking part of your custom Flowzone action would be to use sed to replace the image string in the existing `action.yml` before running the action.

eg.
```yml
run: |
  sed -r 's|(image: docker://ghcr\.io/balena-io/deploy-to-balena-action:).+$|\1${{ github.event.pull_request.head.sha }} \
    -i action.yml
```

**Comment by maggie0002** [comment]
I think then it would need to commit that change to be able to access the action.yml file to be able to do the test deploy? Could have a look to see if we can reference a local action.yml without having to commit it, but was aiming to keep it as close to how it deploys normally.

**Comment by klutchell** [comment]
We definitely don't want to commit the change, you should be able to call a local action with `uses: ./action.yml`

eg. https://github.com/product-os/flowzone/blob/master/.github/workflows/tests.yml#L17

**Comment by maggie0002** [comment]
> uses: ./action.yml

Didn't seem to like specifying the action.yml, but specifying the context `./` works ok, finding the action.yml as its default.

**Comment by klutchell** [comment]
> > uses: ./action.yml
> 
> Didn't seem to like specifying the action.yml, but specifying the context `./` works ok, finding the action.yml as its default.

Maybe using the full .yml path only works for reusable workflows. Either way, what you have seems to be correct!

File: action.yml [hunk 1]
```diff
@@ -48,7 +48,7 @@ inputs:
     required: false
     default: true
   registry_secrets:
-    description: Image registry credentrials (Balena secrets.json)
+    description: Image registry credentials (Balena secrets.json)
```
Comment (klutchell) [inline action.yml:51]: I would prefer that fixing typos or whitespace be in a separate commit.

Some reasoning is that reverting, or cherry-picking, functional commits is much easier when it doesn't include unrelated changes.

It's also excellent practice for rebasing and splitting commits!

File: .github/workflows/flowzone.yml [hunk 2]
```diff
@@ -18,3 +18,23 @@ jobs:
...
+    runs-on: ubuntu-latest
+    needs: flowzone
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Replace the build image and Dockerfile for testing
+        run: |
+          yq e -i '.runs.image = "docker://ghcr.io/balena-io/deploy-to-balena-action:${{ github.event.pull_request.head.sha }}"' \
+            action.yml
+          echo FROM busybox > Dockerfile
```
Comment (klutchell) [inline .github/workflows/flowzone.yml:33]: Might be worth a comment here to explain why the Dockerfile is being overwritten.

File: .github/workflows/flowzone.yml [hunk 3]
```diff
@@ -18,3 +18,25 @@ jobs:
     with:
       docker_images: |
         ghcr.io/balena-io/deploy-to-balena-action
+
+  test-deploy-to-cloud:
+    name: Test deploy to balenaCloud
+    runs-on: ubuntu-latest
+    needs: flowzone
```
Comment (klutchell) [inline .github/workflows/flowzone.yml:25]: I think we would want a condition here to make sure this is not a closed event. Running on a closed event wouldn't be an issue but in general tests that run after a merge or close are pretty useless.
```suggestion
    needs: flowzone
    if: github.event_name == 'pull_request' && github.event.action != 'closed'
```

Comment (maggie0002) [reply .github/workflows/flowzone.yml:25]: I think the action already only runs on pull_requests?

```
on:
  pull_request:
```

But perhaps just the `github.event.action != 'closed'`?

Comment (klutchell) [reply .github/workflows/flowzone.yml:25]: That's correct, I just tend to err on the side of catching unpredicted use cases, like someone adding a new trigger without updating the if conditions. Just checking for not closed is fine.

**Review by klutchell — APPROVED** [review]

**Comment by klutchell** [comment]
lgtm
